STM32F1/libraries/A_STM32_Examples/examples/Sensors/HardTimerAsEncoder/HardTimerAsEncoder.ino

```diff
@@ -123,35 +84,78 @@ EENCODER SIMULATION PART.
  * if received 3 - Mode 3 (Counts on both channels)
  * if received 4 - Change prescaler to 4
  * if received 0 - change prescaler to 1
- * if received - - Increase Speed
- * if received + - Decrease Speed
+ * if received - - Decrease Speed
+ * if received + - Increase Speed
 */
 
 //take care of comms...
   if (Serial.available() > 0) {
-    received = Serial.read();
+    char received = Serial.read();
     if (received == 'F' || received == 'R') dir = received; //direction. Forward or Reverse.
     if (received == '1') timer.setEdgeCounting(TIMER_SMCR_SMS_ENCODER1); //count only the pulses from input 1
     if (received == '2') timer.setEdgeCounting(TIMER_SMCR_SMS_ENCODER2); //count only the pulses from input 2
     if (received == '3') timer.setEdgeCounting(TIMER_SMCR_SMS_ENCODER3); //count on both channels (default of the lib).
     if (received == '4') timer.setPrescaleFactor(4); //only updates on overflow, so you need to wait for an overflow. Not really used...  
     if (received == '0') timer.setPrescaleFactor(1); //only updates on overflow, so you need to wait for an overflow. 
-    if (received == '-') freq+=50; //increase speed.
+    if (received == '-') updatePeriod+=periodDelta;  //decrease speed.
     if (received == '+') {
-      freq-=50;
-      if (freq <= 0) freq = 100; //smallest is 10 ms. 
+      if (updatePeriod <= periodDelta) updatePeriod = 50; //smallest is 50 ms. 
+      else updatePeriod-=periodDelta;
     }
   }
   
 //simulate encoder pulses.   
-  if ( millis() - time >= freq) { 
-    time = millis(); //prepare next
+  
+  if ( millis() - lastSimulationAt >= updatePeriod) { 
+    lastSimulationAt = millis(); //prepare next
     
     if (dir == 'F')  mode++;
-    if (dir == 'R')  mode --;
-    
-    digitalWrite(4, (states[mode%4] & 0x01));
-    digitalWrite(5, (states[mode%4] >> 1));
+    if (dir == 'R')  mode--;
+
+    simulateStep(mode);
   }
+}
+
+void setup() {
+  Serial.begin(115200);
+  while(!Serial)
+    ;
+  //define the Timer channels as inputs. 
+  pinMode(input_a, INPUT_PULLUP);  //channel A
+  pinMode(input_b, INPUT_PULLUP);  //channel B
+
+//configure timer as encoder
+  //set mode, the channel is not used when in this mode.
+  // You must, however, use a valid channel number.
+  timer.setMode(TIMER_CH1, TIMER_ENCODER); 
+  timer.pause(); //stop... 
+  timer.setPrescaleFactor(1); //normal for encoder to have the lowest or no prescaler. 
+  timer.setOverflow(PPR);    //use this to match the number of pulse per revolution of the encoder. Most industrial use 1024 single channel steps. 
+  timer.setCount(0);          //reset the counter. 
+  timer.setEdgeCounting(TIMER_SMCR_SMS_ENCODER3); //or TIMER_SMCR_SMS_ENCODER1 or TIMER_SMCR_SMS_ENCODER2. This uses both channels to count and ascertain direction. 
+  timer.attachInterrupt(0, overflowInterrupt); //channel must be 0 here.  
```

Comment:
Shouldn't be a `timer.refresh()` here, before `attachInterrupt()`?

Reply:
YES! Thank you. that solved the ARR problem